### PR TITLE
WIP: OWLS-95596 - WKO 4.0 Elasticity and Autoscaling

### PR DESCRIPTION
**Pull request by lennyphan**
This contains implementation changes for scaling using Cluster Resource.  This is PR is still a WIP and is so that everyone can begin to see if we're moving in the direction we want for handling Cluster Resources and at this point contains the following changes:

- Define Cluster Resource CRD
- Version webhook changes for converting v8 -> v9 - v8 fidelity.
- Renamed the 'Cluster' element representing cluster configuration to ClusterSpec to:
       -   Cluster object now represents the Cluster Resource model.
       -   ClusterSpec name aligns with the specification of the cluster configuration attributes, same as DomainSpec for 
            Domain configuration attributes..
- Moved effective configuration and resource validation from Domain to DomainPresenceInfo since Cluster Resource objects will be stored in the DomainPresenceInfo and the effective configuration will need to take into account the ClusterSpec configuration from Cluster Resources.
- Handle watch events for Cluster Resources in a namespace.
- Modified unit tests to access Cluster configuration through DomainPresenceInfo.
- CallBuilder/WebLogic API changes to create, list and patch Cluster Resources.

### CODE REVIEW

File: common/src/main/java/oracle/kubernetes/common/ClusterCustomResourceHelper.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import java.util.List;
+import java.util.Map;
+
+import io.kubernetes.client.openapi.ApiException;
+
+public interface ClusterCustomResourceHelper {
+  void createClusterResource(Map<String, Object> clusterSpec, Map<String, Object> domain) throws ApiException;
+
+  Map<String, Object> getDeployedClusterResources(String namespace, String domainUid);
```
[the remark this answers is not in the document]
Comment (lennyphan) [reply common/src/main/java/oracle/kubernetes/common/ClusterCustomResourceHelper.java:14]: Yes, just listing the Cluster resources.  I will change the name.

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/ClusterSpec.java [hunk 2]
```diff
@@ -3,67 +3,293 @@
...
-   * Returns the labels applied to the cluster service.
-   *
-   * @return a map of labels
-   */
+/**
+ * An element representing a cluster in the domain configuration.
+ *
+ * @since 2.0
+ */
+public class ClusterSpec extends BaseConfiguration implements Comparable<ClusterSpec> {
```
Comment (rjeberhard) [inline operator/src/main/java/oracle/kubernetes/weblogic/domain/model/ClusterSpec.java:27]: I was expecting to see a field that referenced the Domain that this Cluster belongs to... or, is that only specified on the Domain side (i.e. Domain lists Cluster resources associated with the Domain)?

File: common/src/main/java/oracle/kubernetes/common/utils/SchemaConversionUtils.java [hunk 3]
```diff
@@ -104,6 +120,128 @@ public String convertDomainSchema(String domainYaml) {
     return yaml.dump(convertDomainSchema((Map<String, Object>) domain));
   }
 
+  private void convertClustersToResources(ClusterCustomResourceHelper clusterCustomResourceHelper,
+      Map<String, Object> domain, String desiredAPIVersion) {
+    try {
+      if (clusterCustomResourceHelper != null) {
```
Comment (rjeberhard) [inline common/src/main/java/oracle/kubernetes/common/utils/SchemaConversionUtils.java:126]: Can this be null?

File: common/src/main/java/oracle/kubernetes/common/utils/SchemaConversionUtils.java [hunk 4]
```diff
@@ -104,6 +120,128 @@ public String convertDomainSchema(String domainYaml) {
...
+    if (!clusterSpecs.isEmpty()) {
+      List<String> clusterResources = clusterCustomResourceHelper
+          .getNamesOfDeployedClusterResources(getNamespace(domain), getDomainUid(domain));
+      Iterator<Object> itr = clusterSpecs.iterator();
+      while (itr.hasNext()) {
+        Map<String, Object> clusterSpec = (Map<String, Object>) itr.next();
+        String clusterName = (String) clusterSpec.get("clusterName");
+        if (!clusterResources.contains(clusterName)) {
+          // Create Cluster Resource
+          clusterCustomResourceHelper.createClusterResource(clusterSpec, domain);
```
Comment (rjeberhard) [inline common/src/main/java/oracle/kubernetes/common/utils/SchemaConversionUtils.java:154]: There are likely several use cases here, such as where a Cluster resource already exists and is the same or different than the resource the code wants to create.

File: common/src/main/java/oracle/kubernetes/common/utils/SchemaConversionUtils.java [hunk 5]
```diff
@@ -104,6 +120,128 @@ public String convertDomainSchema(String domainYaml) {
...
+    }
+  }
+
+  private void addClusterResourceReference(Map<String, Object> domainSpec, String clusterName) {
+    Optional.ofNullable(getClusterResourceReferences(domainSpec)).map(c -> c.add(clusterName));
+  }
+
+  protected List<String> getClusterResourceReferences(Map<String, Object> domainSpec) {
+    List<String> clusterResourceReferences =
+        (List<String>) domainSpec.get(CLUSTER_RESOURCE_REFERENCES);
```
Comment (rjeberhard) [inline common/src/main/java/oracle/kubernetes/common/utils/SchemaConversionUtils.java:172]: "cluster-resource-references" is a mouthful... I'm wondering if we can just still use "clusters" but have the content now be the list of references thereby highlighting that the schema is mostly unchanged but just that this content has been promoted to its own resource.

File: operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java [hunk 6]
```diff
@@ -624,6 +634,73 @@ private void handleDeletedDomain(Domain domain) {
...
+
+  private void handleAddedCluster(Cluster cluster) {
+    System.out.println("DomainProcessorImp.handleAddedCluster cluster: " + cluster);
+    DomainPresenceInfo cachedInfo = getExistingDomainPresenceInfo(cluster.getNamespace(), cluster.getDomainUid());
+    System.out.println("DomainProcessorImp.handleAddedCluster cachedInfo: " + cachedInfo);
+    if (cachedInfo == null) {
+      return;
+    }
+
+    //LOGGER.info(MessageKeys.WATCH_DOMAIN, domain.getDomainUid());
```
Comment (rjeberhard) [inline operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java:667]: Find and delete commented out code.

Comment (rjeberhard) [reply operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java:667]: Oh, wait... This seems like code that should be uncommented and should initiate domain processing.

File: operator/src/main/java/oracle/kubernetes/operator/KubernetesConstants.java [hunk 7]
```diff
@@ -22,6 +22,14 @@ public interface KubernetesConstants {
   String DOMAIN_VERSION = "v9";
   String OLD_DOMAIN_VERSION = "v8";
 
+  String CLUSTERS_CRD_NAME = "clusters.weblogic.oracle";
```
Comment (rjeberhard) [inline operator/src/main/java/oracle/kubernetes/operator/KubernetesConstants.java:25]: What is this used for? I was going to say that I would like the Clusters resource to be in the same group as the Domain resource -- I think that you are doing this based on what I see below.

File: operator/src/main/java/oracle/kubernetes/operator/KubernetesConstants.java [hunk 8]
```diff
@@ -22,6 +22,14 @@ public interface KubernetesConstants {
...
+  String CLUSTER = "Cluster";
+  String CLUSTER_PLURAL = "clusters";
+  String CLUSTER_SINGULAR = "cluster";
+  String CLUSTER_SHORT = "clu";
+  String CLUSTER_VERSION = "v9";
```
Comment (rjeberhard) [inline operator/src/main/java/oracle/kubernetes/operator/KubernetesConstants.java:30]: This is Cluster v1 :)

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/Domain.java [hunk 9]
```diff
@@ -712,18 +645,18 @@ public String getLivenessProbeCustomScript() {
     return emptyToNull(spec.getLivenessProbeCustomScript());
   }
 
-  public boolean isShuttingDown() {
-    return getEffectiveConfigurationFactory().isShuttingDown();
-  }
+  //public boolean isShuttingDown() {
```
Comment (rjeberhard) [inline operator/src/main/java/oracle/kubernetes/weblogic/domain/model/Domain.java:648]: Okay, this is actual commented out code that I think that you need to delete :)

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java [hunk 10]
```diff
@@ -414,19 +413,64 @@ DomainSpec withFluentdConfiguration(boolean watchIntrospectorLog, String credent
...
+   * The list of names of cluster custom resources the operator should reference for cluster
+   * configuration elements.
+   *
+   * @since 4.0
+   */
+  @Description("List of names of Cluster Custom Resources the Operator expects to be deployed in the "
+      + "same namespace that the domain is deployed.  A reference to a cluster custom resource "
+      + " in order for the cluster elements to be applied.")
+  @SerializedName("cluster-resource-references")
+  protected final List<String> clusterResourceReferences = new ArrayList<>();
```
Comment (rjeberhard) [inline operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java:428]: I think this should be List<V1LocalObjectReference>

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java [hunk 11]
```diff
@@ -414,19 +413,64 @@ DomainSpec withFluentdConfiguration(boolean watchIntrospectorLog, String credent
       + "environment variables, additional Pod content, and the ability to explicitly start, stop, or restart "
       + "cluster members. The `clusterName` field of each entry must match a cluster that already exists in the "
       + "WebLogic domain configuration.")
-  protected final List<Cluster> clusters = new ArrayList<>();
+  protected final List<ClusterSpec> clusters = new ArrayList<>();
```
Comment (rjeberhard) [inline operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java:416]: What is this used for? Will it end up in the schema?

File: common/src/main/java/oracle/kubernetes/common/utils/SchemaConversionUtils.java (hunk 5 above)
Comment (lennyphan) [reply common/src/main/java/oracle/kubernetes/common/utils/SchemaConversionUtils.java:172]: Yes, I that should be able to be done.

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/ClusterSpec.java (hunk 2 above)
Comment (lennyphan) [reply operator/src/main/java/oracle/kubernetes/weblogic/domain/model/ClusterSpec.java:27]: At the moment, the associated Domain is referenced as a label in the Cluster Resource:

e.g.
apiVersion: "weblogic.oracle/v1"
kind: Cluster
metadata:
  name: cluster-1
  namespace: sample-domain1-ns
  labels:
    weblogic.domainUID: sample-domain1

I didn't add or modify the attributes of the Cluster spec from the original.  I've been identifying the associated domain by using the label.

File: operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java (hunk 6 above)
Comment (lennyphan) [reply operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java:667]: Sorry, I haven't cleaned up as I'm currently working on refactoring and abstracting out the conditions for initiating domain processing for either Domain and/or Cluster events.

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java (hunk 11 above)
Comment (lennyphan) [reply operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java:416]: I've not  removed the Clusters configuration from the DomainSpec as much of our unit tests are still relying on it.  I didn't modify the unit tests yet as I had just moved most of the effective configuration from the Domain to DomainPresenceInfo and those unit test are validating that the move was fine.

**Comment by tbarnes-us** [comment]
Suggestion: Consider adding the name of the associated Cluster Resource to the cluster status.

File: operator/src/main/java/oracle/kubernetes/operator/KubernetesConstants.java (hunk 7 above)
Comment (lennyphan) [reply operator/src/main/java/oracle/kubernetes/operator/KubernetesConstants.java:25]: This is the name of the CRD for Cluster kind.  Cluster Resources are still part of the same group ans the Domain resource.

bash-4.2$ kubectl get crds
NAME                       CREATED AT
clusters.weblogic.oracle   2022-05-19T17:29:48Z

**Pull request closed without merging**